source/mesh_refinement/nonadiabatic_temperature_threshold.cc

```diff
@@ -0,0 +1,156 @@
+/*
+  Copyright (C) 2015 - 2023 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/mesh_refinement/nonadiabatic_temperature_threshold.h>
+#include <aspect/adiabatic_conditions/interface.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+namespace aspect
+{
+  namespace MeshRefinement
+  {
+    template <int dim>
+    void
+    NonadiabaticTemperatureThreshold<dim>::tag_additional_cells () const
+    {
+      // tag_additional_cells is executed before the equations are solved
+      // for the very first time. If we do not have the finite element, we
+      // do not have the temperature and just do nothing in this plugin.
+      if (this->get_dof_handler().n_locally_owned_dofs() == 0)
+        return;
+
+      const Quadrature<dim> quadrature(this->get_fe().base_element(this->introspection().base_elements.temperature).get_unit_support_points());
+      FEValues<dim> fe_values (this->get_mapping(),
+                               this->get_fe(),
+                               quadrature,
+                               update_quadrature_points | update_values);
+
+      std::vector<double> temperature_values (quadrature.size());
+
+      for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+        if (cell->is_locally_owned())
+          {
+            bool refine = false;
+
+            fe_values.reinit(cell);
+            fe_values[this->introspection().extractors.temperature].get_function_values (this->get_solution(),
+                                                                                         temperature_values);
+
+            // if the nonadiabatic temperature exceeds the threshold, cell is marked for refinement
+            for (unsigned int j=0; j<this->get_fe().base_element(this->introspection().base_elements.temperature).dofs_per_cell; ++j)
```

Comment:
Can you compute the loop bound once outside the loop?

Reply:
Like this?

The previous version is how it's done in a couple of other plugins (where I copied this from). Should I go through those and change it as well? I think it would make sense to do this the same way everywhere.

Reply:
I think you can move it outside the cell loop as well. If you have time to do it, feel free to look at other places as well.

Reply:
Alright, I've moved it out of the cell loop as well, and I'll make a follow-up PR for the other plugins.